epochStart/bootstrap/shardStorageHandler.go

```diff
@@ -882,34 +882,70 @@ func (ssh *shardStorageHandler) saveTriggerRegistry(components *ComponentsNeeded
 	return bootstrapKey, nil
 }
 
-func getMiniBlockHeadersForDest(metaBlock *block.MetaBlock, destId uint32) map[string]block.MiniBlockHeader {
+func getMiniBlockHeadersForDest(metaBlock data.MetaHeaderHandler, destId uint32) map[string]block.MiniBlockHeader {
 	hashDst := make(map[string]block.MiniBlockHeader)
-	for i := 0; i < len(metaBlock.ShardInfo); i++ {
-		if metaBlock.ShardInfo[i].ShardID == destId {
+	for i := 0; i < len(metaBlock.GetShardInfoHandlers()); i++ {
+		if metaBlock.GetShardInfoHandlers()[i].GetShardID() == destId {
 			continue
 		}
 
-		for _, val := range metaBlock.ShardInfo[i].ShardMiniBlockHeaders {
-			isCrossShardDestMe := val.ReceiverShardID == destId && val.SenderShardID != destId
+		for _, val := range metaBlock.GetShardInfoHandlers()[i].GetShardMiniBlockHeaderHandlers() {
+			isCrossShardDestMe := val.GetReceiverShardID() == destId && val.GetSenderShardID() != destId
 			if !isCrossShardDestMe {
 				continue
 			}
 
-			hashDst[string(val.Hash)] = val
+			miniBlockHeader, ok := val.(*block.MiniBlockHeader)
+			if !ok {
+				log.Warn("wrong type assertion for mini block header handler", "err", epochStart.ErrWrongTypeAssertion)
+				continue
+			}
+			hashDst[string(val.GetHash())] = *miniBlockHeader
 		}
 	}
 
-	for _, val := range metaBlock.MiniBlockHeaders {
-		isCrossShardDestMe := (val.ReceiverShardID == destId || val.ReceiverShardID == core.AllShardId) && val.SenderShardID != destId
+	miniBlockHandlers := getMetaHeaderMiniBlockHandlersFromExecutionResults(metaBlock)
+
+	for _, val := range miniBlockHandlers {
+		isCrossShardDestMe := (val.GetReceiverShardID() == destId || val.GetReceiverShardID() == core.AllShardId) && val.GetSenderShardID() != destId
 		if !isCrossShardDestMe {
 			continue
 		}
-		hashDst[string(val.Hash)] = val
+
+		miniBlockHeader, ok := val.(*block.MiniBlockHeader)
+		if !ok {
+			log.Warn("wrong type assertion for mini block header handler", "err", epochStart.ErrWrongTypeAssertion)
+			continue
+		}
+		hashDst[string(val.GetHash())] = *miniBlockHeader
 	}
 
 	return hashDst
 }
 
+func getMetaHeaderMiniBlockHandlersFromExecutionResults(
+	metaBlock data.MetaHeaderHandler,
+) []data.MiniBlockHeaderHandler {
+	miniBlockHeaderHandlers := metaBlock.GetMiniBlockHeaderHandlers()
+	if !metaBlock.IsHeaderV3() {
+		return miniBlockHeaderHandlers
+	}
+
+	lastExecResult, err := common.GetLastBaseExecutionResultHandler(metaBlock)
+	if err != nil {
+		log.Warn("failed to get last execution result from meta header v3", "err", err)
+		return miniBlockHeaderHandlers
```

Comment:
I think you should return nil instead, as returning the proposed miniblocks from a meta v3 block, will cause marking them as processed even though they are not, right?

Reply:
right, returned nil